gpuRIR/extensions/filtering.py

```diff
@@ -0,0 +1,262 @@
+import numpy as np
+import numpy.matlib
+import matplotlib.pyplot as plt
+from math import ceil
+from scipy.io import wavfile
+import time
+
+from gpuRIR.extensions.filters.filter import Filter
+from gpuRIR.extensions.filters.characteristic_filter import CharacteristicFilter
+from gpuRIR.extensions.filters.air_absorption_bandpass import AirAbsBandpass
+from gpuRIR.extensions.filters.air_absorption_stft import AirAbsSTFT
+from gpuRIR.extensions.filters.linear_filter import LinearFilter
+import gpuRIR.extensions.filters.characteristic_models as cm
+import gpuRIR.extensions.filters.air_absorption_calculation as aa
+from gpuRIR.extensions.wall_absorption.materials import Materials as mat
+import gpuRIR.extensions.wall_absorption.freq_dep_abs_coeff as fdac
+from gpuRIR.extensions.create_spectrogram import create_spectrogram_from_data
+import gpuRIR.extensions.room_parameters as rp
+from gpuRIR.extensions.generate_RIR import generate_RIR
+
+
+""" Generates an impulse response WAV file (IR) with optional filters.
+Example usage: Convolving (reverberating) an audio signal in an impulse response loader plug-in like Space Designer in Logic Pro X.
+"""
+
+
+def mono_adaptive_gain(source, bit_depth, ceiling):
+    ''' Increases amplitude (loudness) to defined ceiling. Operates in an adaptive manner to prevent clipping.
+
+    Parameters:
+    ----------- 
+    source : 2D ndarray
+        Sound data to process.
+    bit_depth : int
+        Bit depth of source sound data.
+    ceiling : int
+        Maximum loudness (relative dB, e.g. -1dB) the sound data should be amplified to
```

Comment:
Maybe also say additionally. "This way, the loudest sample will result being to ceiling in dB (eg. -1dB)